File: src/leetCode/Ex31.java
package leetCode;

public class Ex31 {
    public void nextPermutation(int[] nums) {
        int len = nums.length;
        if(len<2){
            return;
        }
        int l = -1;
        int r = -1;
        
        int pre = nums[len-1];
        for (int i = len-2; i >= 0; i--) {
            if(nums[i]<pre){
                l = i;
                break;
            }else{
                pre = nums[i];
            }
        }
        if(l==-1){
            for (int i = 0; i < len&& i<len-i-1; i++) {
                swap(nums, i, len-i-1);
            }
            return;
        }
        for (int i = len-1; i > l; i--) {
            if(nums[i]>nums[l]){
                r = i;
                break;
            }
        }
        //swap l and r;
        swap(nums, l, r);
        for (int i = l+1; i < len&& i<len-i+l; i++) {
            swap(nums, i, len-i+l);
        }

    }
    private void swap(int[] nums, int i, int j){
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}

File: src/LeetCode/ex31.kt
package LeetCode

fun nextPermutation(nums: IntArray) {
    var r = nums.size-2
    while (r>=0&&nums[r]>=nums[r+1]) r--
    when (r) {
        -1 -> {
            verse(nums, 0, nums.size - 1)
        }
        nums.size-2-> verse(nums, r, r + 1)
        else -> {
            var s = r + 1
            while (s < nums.size && nums[r] < nums[s]) s++
            wap(nums, r, s - 1)
            verse(nums, r + 1, nums.size - 1)
        }
    }
}

private fun wap(nums: IntArray,i:Int,j:Int){
    val temp = nums[i]
    nums[i]=nums[j]
    nums[j]=temp
}

private fun verse(nums: IntArray,start:Int,end:Int){
    val len = end-start+1
    (0 until len/2).forEach { i -> wap(nums, i = i + start, j = end - i) }
}

fun main(args: Array<String>) {
    val nums = intArrayOf(1,3,2)
    nextPermutation(nums)
    nums.forEach { print(""+it+" ") }
}